### PR TITLE
Refactor building.md

### PR DESCRIPTION
**Pull request by cramja**
* Added table of content
* updated building instructions
* Removed some redundancy
* Moved bonus sections into the appendix

### CODE REVIEW

**Comment by cramja** [comment]
> The job exceeded the maximum time limit for jobs, and has been terminated.

File: BUILDING.md [hunk 1]
```diff
@@ -1,60 +1,71 @@
...
-used).
+**Short Version**
 
-### Vagrant
+```sh
+cd third_party
+./download_and_patch_prerequisites.sh
+cd ../build
+cmake ..
+make -j4 quickstep_cli_shell
```
Comment (zuyu) [inline BUILDING.md]: I'd suggest to remove `-j4` here, as it does not apply to everyone, and we have mentioned the parallel build jobs later.

Comment (cramja) [inline BUILDING.md]: sure

**Comment by cramja** [comment]
changed and closed

**Pull request closed without merging**